doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,201 @@
+|:------------------------- |:--------- |:--------- |:--------- |:--------- |:--------- |:--------- |
+| Software Timers           | ✓         | ✓         | ✓         | ✓         | ✓         | ✗         |
+| Periodic Software Timers  | ✓         | ✓         | ✓         | ● (2)     | ✓         | ✗         |
+| Context of Timer Callback | Thread    | ?         | ?         | Both? (3) | ?         | n.a.      |
+| System Time               | ✓         | ✓         | ✓         | ✓         | ✓         | ✓         |
+| Delays                    | ✓         | ● (1)     | ✓         | ● (4)     | ✓ (5)     | ✓         |
+
+✓ = 1st level API support, ● = supported by combining APIs, ✗ = no support
+
+1. Threads can wait until a software timer fires. Software timers accept NULL
+   as empty callback. Combined, this can be used to delay execution.
+2. Contiki provides with `ctimer_reset()`/`etimer_reset()` which restarts the
+   timer from the previous expiration time, allowing to easily achieve
+   periodic callbacks.
+3. The Rtimer library seems to run the callbacks in IRQ context and the Ctimer
+   library in thread context.
+4. The Etimer library posts events on timer expiration and a thread can block
+   until a specific event has been posted. Combined, this can be used to
+   delay execution
+5. Only available when using TOSTThreads
+
+# 2. Requirements
+
+The following lists of requirements consist of qualitative and quantitative
+requirements. For the quantitative requirements, any improvement will enable
+more scenarios to use the timer API. Thus, no meaningful lower bounds can be
+specified. Instead, a set of benchmarks will be developed that allow to
+continuously improve the performance and to detect any possible regressions of
+future changes.
+
+## Feature Requirements
+
+- Software Timers
+    - Upper bound for number of timers only given by available resources
+    - APIs to set and abort timers
+    - Support for periodic timers (support via explicit calls to some
+      `timer_reset()` like in Contiki is fine)
+    - Long term low resolution software timers that play well with power management
+    - Short term high resolution software timers
+- Delay of program execution
+    - API to block calling thread for long periods and allows other threads to run
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable units
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time
```

Comment:
I'm pretty sure that at least for list-based timers, e.g. using a mutex for the list operations is definitely possible, but harder (more complex) than just disabling IRQs. It might also imply a tradeoff not only in code simplicity but also runtime / code size. So, I'm not happy with the "must" here.

Reply:
The "short periods of time" is deliberately a bit vague to give the implementation some freedom in this regard. But there is clearly an advantage in having IRQs only disabled for a bounded time for real time tasks. This bound could also be an `O(n)` with `n` being the worst cast number of software timers used concurrently; any specific use case will have a specific fixed `n` there, so that the worst case time of a timer implementation for that specific use case can be calculated and real timer properties can be verified.

Reply:
How about this? (See latest fixup)

Reply:
Better. :)

I think `O(n)` hardly qualifies as "real-time", but as we're are talking "requirements", "real-time" would certainly be nice-to-have, but the trade-offs involved might make it infeasible as default implementation, if we'd go stricter than `O(n)`.